### DISCUSSION TITLE
minimal cmake build instructions? - Readme seems out of date.

### DISCUSSION THREAD

**Discussion by sequoia-hope** [earlephilhower/arduino-pico#1235] [General]
I am trying to figure out the minimal setup to build a sketch with cmake. In the readme, the only mention of building with cmake goes to a two year old Toms Hardware article, and the people in the comments of that article mention that the tutorial is completely broken. Also that tutorial makes no mention of cmake.

Anyway I very much appreciate this library. I am trying to merge my application with a [serial bootloader](https://github.com/rhulme/pico-flashloader) someone has made that uses cmake. Any instructions on a basic cmake build setup would be greatly appreciated. Thanks!

**Comment by earlephilhower** [comment]
Sorry, I think there's a bit of a misunderstanding here.  The `cmake` info was for building plain C/C++ apps, not building Arduino apps.  Back when the Pico-SDK came out, getting the SDK up and running was a pain and the Tom's Hardware article was a gret help.  Now, I think there's a more developed infrastructure (and a 1-click Windows installer IIRC).

You can use the `arduino-cli` or Platform.IO to compile an Arduino project from the command line or in a `Cmakefile` target.  There's no Cmakefile recipe for building an Arduino app, to the best of my knowledge.

I'll move this to a discussion since it's more "how do I" vs. a core problem.

**Comment by sequoia-hope** [comment]
Ah, I understand, thank you for the clarification!

**Comment by daveythacher** [comment]
I was considering adding support for copy to RAM configurations. In summary there is a linker workflow which likely could be modified to support a boot loader. (I do not really recommend that one unless you know what your doing.)

The documentation does not cover the internals of the linker script and configuration files. I think I may have gotten around that, but could not get a clear way to test my changes. The package logic did not work as it generated a 404. There is one page which mostly explains certain things, took a few reads and guessing to make sense of it.

The way this works as I understand it, is there is a configuration file used by the Arduino logic. This contains commands to call a linker script via tool chain. CMake is completely removed by the time the package is created. A library is created which gets linked in. More than likely you can get what you want with this. Its kind of a neat linker trick.

This linker script is installed by the board manager. You can modify it within the local folder. (I would not do this unless you know what your doing.)

**Reply by daveythacher** [reply]
My abandoned pull request is #1225

**Comment by sequoia-hope** [comment]
Thank you! Actually it turns out that the features I need are already implemented in this software! I got all the way to the point of integrating another serial bootloader when I noticed notes about "OTA" updates in the linker script, which led me to find this:
https://github.com/earlephilhower/arduino-pico/blob/master/libraries/PicoOTA/examples/OTAfromFile/OTAfromFile.ino

I was able to use that to do all the updates for my system!

**Reply by daveythacher** [reply]
I have concerns for OTA based on this: 
> The bootloader is built here into an .ELF, without boot2.S (which will come from the main app), configured to copy itself into RAM (so that it can update itself), and included in the main applications. Exactly 12KB for all sketches is consumed by this OTA bootloader.

...
> Should a power failure happen, as long as it was not in the middle of writing a new OTA bootloader, it should simply begin copying the same program from scratch.

> When the copy is completed, the command file's contents are erased so that on a reboot it won't attempt to write the same firmware over and over. It then reboots the chip (and re-runs the potentially new bootloader).

This would allow updating without the OTA library which could brick the device. Also if a failure occurs in the right spot it will in fact fail into a bricked state. If the new image is found to be corrupt the recovery could be improbable. Boot loader solves this by fall back image or being the fall back image itself. (The RP2040 comes with a fall back with the internal ROM via USB.)

Granted the model used here is actually desirable for many people, however it does come with risks.

**Reply by sequoia-hope** [reply]
Ah this makes sense, thanks for the advice. For our case, we are updating a robot with eight rp2040s on it. This bootloader is a convenience so we do not have to hook up a laptop and USB cable eight times to do code updates. But if we do experience a failure we can recover using the boot switch and built in bootloader over USB. We will have to try it and see if we run in to any kinds of trouble.

At least we can ensure we always build with the appropriate library, because we control the environment (this is not a general purpose board for hobbyists).

**Reply by earlephilhower** [reply]
FWIW, the OTA is not rewritten, nor the boot2, unless the update file has a different version.  So assuming there is no Ota.CPP update (hasn't been one yet) I don't think there is a failure mode that way.  The copy routine always checks the existing flash before erasing and rewriting.

**Comment by myklemykle** [comment]
I have been trying to bring up [this Arduino CMake project](https://github.com/technyon/Arduino-CMake-Toolchain), which apparently works with both AVR and ESP32, and is intended to work with any core. But I might be the first person to try it on the rp2040 core. It's not quite working. (issues [here](https://github.com/technyon/Arduino-CMake-Toolchain/issues/3)) .

I'm afraid I have no idea how this is all supposed to work behind the scenes. Is there an arduino+cmake API that cmake scripts are supposed to use to find paths & generate object files in any Arduino core?

It would sure be brilliant if it worked. arduino-cli is super-slow, and also i'm starting to need different versions of libraries in different projects.

**Reply by daveythacher** [reply]
Arduino has been dying for about a decade now. The IDE was unstable then and its much worse now. The installer does not even appear to work in some cases.

What this Arduino Core does with the SDK is converts it into a static library. This can then be consumed by a make process. CMake is not used by Arduino. Therefore AVR core compatibility should still work, unless people made non-standard Arduino hooks. 

Arduino is a beautiful framework at the end of the day with only so much support. Conflict and opinionated people undermined Arduino to what it is today.

**Reply by myklemykle** [reply]
Arduino continues to be a really great multi-platform ecosystem full of open-source packages that are well-tested, well-documented and well-supported. The goal of "design it for beginners" has resulted in a lot of very clear & well thought out APIs. I have looked at some vendor-provided library ecosystems & they made me cringe.

I don't use the IDE at all. My current irritation is with the CLI. The Arduino build process starts with fake-compiling the sketch over and over in order to discover unresolved #include directives, in order to do its weird library discovery/selection process. This is super slow, it gets slower the more libraries you use, and the results are never cached even though they don't change. 80% of my compile cycle, timewise, is just that same library resolution process over and over and over. I've grepped the doc and I find no way to assist or override or hardwire it.  So I'm trying to replace all that with cmake, or just a makefile.

**Reply by daveythacher** [reply]
The standard Arduino process is Make with compiler.  This project produces a static library which is added to build via boards.txt and platform.txt. It does not produce a CMake or Make project. Arduino does not support real dependencies and I am not sure they really wanted it. 

What this means is you can bring in this project manually. The Arduino parsers should be available somewhere in the Arduino IDE source code. This is probably a more direct approach to what you want. (Technically you could wrap this in CMake...)

**Comment by myklemykle** [comment]
What is comes down to is various errors in the final link stage, where objects/libraries/classes that clearly exist are not getting found. So really it seems like this is very close to working. I just don't know why the combination of cmake & the rp2040 core ends up looking in the wrong places for object files.

In particular, `cmake` locates the Adafruit_TinyUSB library just fine, but the subsequent `make` still gets compiler errors about missing those classes. (Or else it's looking for some other class I don't understand, that's also Adafruit and USB related?)
